### ISSUE TITLE
DWX_ZeroMQ_Server_v2.0.1_RC8.mq4 mostly crashing

### ISSUE THREAD

**Issue by reppolice** [darwinex/DarwinexLabs#9]
Hi,

I created an 8GB RAM VM for this, which has the latest Windows 10 Enterprise , live MT4 and Python 3.7 32 bit (I suspect with 0mq there is no difference between python 32 bit and 64). The VM itself has given me no weirdness so far. With this new DWX I am getting 99% of the time problems, though they do not seem 100% reproducible. I got no messages at all in "Experts" once or twice, or I got just the Binding message, or I got pricings for a few ticks and then it was gone(not stopped or crashed, it just wasn't printing any more prices). Restarting the computer seems to take a long time to shut MT4 down, and on my current run I only got the Binding messages and decided to turn on the "Live trading" property of the EA, which seems to have crashed my MT4 (Not responding for 10 minutes or more). 

Somewhere in the middle of all this the EA triggered Windows Firewall (do you want to allow...) to which I replied yes, of course. Sometimes turning the firewall off helps with such issues, but I also suspect the server could exit/crash more gracefully. I will probably try on Windows 7 later. From other comments it is clear previous versions run OK for some people, no comments yet on the new one.

Finally it is Friday and the ticks are few but not that few!

Thanks!

**Comment by reppolice** [comment]
Exact same behavior in Windows 7. Hard to believe it is not some kind of bug.

**Comment by TimKabue** [comment]
It has been a while before I used ZeroMQ but I can distinctly remember a situation where I would fail to receive quotes in Python. However, it was always resolved by restarting my MQL4 program and Python program.
I will take a look at my past projects and give more helpful insight.

**Comment by xenu256** [comment]
I has worked previously with MT4 and abandoned it. The platform is unable to handle even hourly quotes without freezing for 100+ instruments, and ticks, obviously, require more resources.

**Comment by integracore2** [comment]
Hi @reppolice,

Thank you for bringing this to our attention, we'll try and reproduce the scenario at our end.

Could you please confirm more precisely the exact sequence of events that lead to the behaviour you experienced, how many symbols were being published, etc

We've tested this with 28 simultaneous symbols publishing from MetaTrader 4 simultaneously with some cases where ticks would drop.

This issue has in fact raised the need to for a test log for this here on GitHub - so thank you for being thorough and raising the flag!

**Comment by reppolice** [comment]
I did not change anything from your README and tutorials, so the symbols were this long list you have in the example file, perhaps 28. I am trying to figure out how to retry in some more predictable and documented way, but meanwhile here's an idea. What if you published a "known working" configuration, in increasing difficulty: 1) a portable MT4 zip with the bridge and example installed. 2) A Windows VM with a working demo 3) A Linux VM with the demo in Wine

**Comment by Steviey** [comment]
I guess it's all about the right socket choosen. I played a lot with the script. 
Meanwhile I defined MT4 as client and R as server (sockets changed). This makes much more sense for me and it works. Even now, I'm able to shut down the server via client deInit-handler and ZeroMQ. Anyway, MT4 and the original concept of the server script could be discussed. Like xenu256, I'm not convinced yet, whether MT4 is the right platform, for such approaches- in general.

**Comment by reppolice** [comment]
> 
> 
> I guess it's all about the right sockets choosen. I played a lot with the script (incl. strategy tester).
> Meanwhile I defined MT4 as client and R as server (sockets changed). This makes much more sense for me and it works! Even now, I'm able to shut down the server via client deInit-handler and ZeroMQ. Anyways, MT4 and the original concept of the server script could be discussed. Like xenu256, I'm not convinced yet, whether MT4 is the right platform, for such approaches- in general.

Are you hinting at MT5, or altogether at greener pastures? I don't know if you went down the "portable MT4" path, the only sensible way to deploy code if you need to move it around between testing and release environments, for example, but certainly the world could use a working template.

**Comment by Steviey** [comment]
I'm hinting at MT4. Not sure if this is worth it.

**Comment by reppolice** [comment]
so MT4 as a client " MT4 is the right platform, for such approaches- in general."? I thought you denounced MT4 altogether for this bridge-automation. Anyway, if it made sense for you, it may make sense for Darwinex devs

**Comment by Steviey** [comment]
@reppolice Yes, perhaps it makes sense for them too. Another question would be, what about the 'one-tick-data' of 28 symbols? Elaborated strategies will use extensive history data per symbol to trade. Maybe one should use this 'one-tick-data' as scanning service. I don't know. There are also to less exceptions. Socket-programming needs an extensive amount of exceptions. They are essential for this kind of programming. And connections have to be terminated correctly. Anything else is a nightmare :-).

**Comment by Steviey** [comment]
@integracore2 Good answer, I'm happy you guys keep care of it and glad to hear that! 
Pleas keep in mind, not everyone prefers Python as the 'logic basis' for algo-tradings.
There is a big R-community in quant environments too. I can code both, but see certain advantages over Python by R. There are also C#-devs out there. It would be nice, to keep the server script open for these platforms. This would provide a much larger user base and more interested, possible members of DrawinX. Thank you in advanced so far.

**Comment by integracore2** [comment]
Hi @Steviey ,

Yes, totally agree regarding additional programming languages being facilitated 👍 

The plan is to make most implementations that we publish on GitHub, available in **Python, R, C#/C++ and Java.** 

The first release of anything will almost always be in Python or R to enable extensive user-testing and incorporating any feedback as necessary - as demonstrated by how this thread has turned out so far, simply fantastic! 😃 

**Regarding the Server EA again** -> I think the solution we discussed earlier about separating the market data publication entirely is the best way forward. 

We've written the necessary code since yesterday to achieve this (in Python for now), and are currently testing out the use-case thoroughly prior to sharing it here.

Will keep you posted! 

And thank you so much for your interest in providing an R version down the line, that's awesome 👏 

We'll make sure its referenced as yours and acknowledged as your contribution by our entire user-base - thank you! 🙏

**Comment by Puhtooie** [comment]
Has there been any consideration to laying a [GraphQL](https://graphql.org/) server over MT4? No matter which language implements it that should effectively open it up to any language that supports the API.
Live data, Trades, and historical data could each have their own tag in the schema.

**Comment by Puhtooie** [comment]
I am extremly interested in adding this library to the following framework:
https://github.com/backtrader/backtrader

Any thoughts on the subject, or individuals also interested in the idea are welcome.
